ui/src/components/EdgePins.tsx

```diff
@@ -0,0 +1,159 @@
+import { Story } from "../types";
+
+export enum EdgeType {
+  Bottom,
+  Right,
+  Top,
+  Left,
+}
+
+function getEdgeType(index) {
+  switch (index) {
+    case 0:
+      return EdgeType.Bottom;
+    case 1:
+      return EdgeType.Right;
+    case 2:
+      return EdgeType.Top;
+    default:
+      return EdgeType.Left;
+  }
+}
+
+function intersects(a, b, c, d, p, q, r, s) {
+  const det = (c - a) * (s - q) - (r - p) * (d - b);
+  if (det === 0) {
+    return false;
+  } else {
+    const lambda = ((s - q) * (r - a) + (p - r) * (s - b)) / det;
+    const gamma = ((b - d) * (r - a) + (c - a) * (s - b)) / det;
+    return 0 < lambda && lambda < 1 && 0 < gamma && gamma < 1;
+  }
+}
+
+function getPointOfIntersection(a, b, c, d, p, q, r, s) {
+  const det = (c - a) * (s - q) - (r - p) * (d - b);
+  const lambda = ((s - q) * (r - a) + (p - r) * (s - b)) / det;
+
+  return L.latLng(a + lambda * (c - a), b + lambda * (d - b));
+}
+
+function angleFromCoordinate(lat1, long1, lat2, long2) {
+  const dLon = long2 - long1;
+
+  const y = Math.sin(dLon) * Math.cos(lat2);
+  const x =
+    Math.cos(lat1) * Math.sin(lat2) -
+    Math.sin(lat1) * Math.cos(lat2) * Math.cos(dLon);
+
+  const brng = (Math.atan2(y, x) * 180) / Math.PI;
+
+  return brng - 80;
+}
+
+export interface EdgePinsProps {
+  stories: Story[];
+  currPosition: LatLng;
+  mapBounds: LatLngBounds;
+}
+
+export function EdgePins({
+  stories,
+  currPosition,
+  mapBounds,
+}: EdgePinsProps): JSX.Element {
+  // determine if the current view has markers
+  if (!mapBounds || !currPosition) return null;
+  let isVisible = true;
+
+  const pinPositions = stories.map((story) => {
+    return { lat: story.latitude, lng: story.longitude };
+  });
+
+  pinPositions.forEach((pos) => {
+    if (mapBounds.contains(L.latLng(pos.lat, pos.lng))) {
+      isVisible = false;
+    }
+  });
+
+  // do not display the edge pins if there are markers in view
+  if (!isVisible) return null;
+
+  // determine the distance from the center to each of the markers
+  const distanceToPos: { [distance: number]: LatLng } = {};
+  pinPositions.forEach((pos) => {
+    const posLatLng = L.latLng(pos.lat, pos.lng);
+    distanceToPos[currPosition.distanceTo(posLatLng)] = posLatLng;
+  });
+
+  //sort the distances and take the first 3
+  const distances = Object.keys(distanceToPos);
+  distances.sort().slice(0, 3);
+
+  // get the closest three pins
+  const closestPins = [];
+  distances.forEach((distance) => {
+    closestPins.push(distanceToPos[distance]);
+  });
+
+  const edgePinsInfo: { pos: LatLng; angle: number; edgeType: EdgeType }[] = [];
+
+  // calculate intersection of the line from the center to the pin
+  // store that point of intersection and the angle of the direction vector in edgePinsInfo
+  const corners = [
+    mapBounds.getSouthWest(),
+    mapBounds.getSouthEast(),
+    mapBounds.getNorthEast(),
+    mapBounds.getNorthWest(),
+  ];
+
+  closestPins.forEach((pin) => {
```

Comment:
Can we pull this out into it's own function? Maybe even add some comments on what it is doing?